File: NetFramework/XHtmlEditor/VB/RichTextEditor/RecentDocumentList.vb
Imports System.IO
Imports System.Collections.Generic
Imports System.Collections.Specialized
Imports System.Text
Imports C1.Win.C1Ribbon
Imports System.Linq

Namespace RichTextEditor
    ''' <summary>
    ''' Class that handles the recently used document list.
    ''' Manages the list that appears in the application's main menu,
    ''' including persistence to the settings file.
    ''' </summary>
    Public Class RecentDocumentList

        Const MAX_ITEMS As Integer = 10
        ' Initializes a new instance of a RecentDocumentList.
        Public Sub New(ByVal docs As List(Of RecentDocumentItem))
            RecentDocuments = docs
        End Sub
        ' add a document to the MRU list
        Public Sub AddDocument(ByVal fileName As String)
            Dim doc As RecentDocumentItem = RecentDocuments.FirstOrDefault(Function(x) x.FileName = fileName)
            If doc IsNot Nothing Then
                doc.ItemDate = DateTime.Now
            ElseIf Not String.IsNullOrEmpty(fileName) Then
                doc = New RecentDocumentItem()
                doc.Text = Path.GetFileName(fileName)
                doc.SubText = Path.GetDirectoryName(fileName)
                doc.ItemDate = DateTime.Now
                RecentDocuments.Add(doc)
            End If

            While RecentDocuments.Count > MAX_ITEMS
                Dim notPinned As IEnumerable(Of RecentDocumentItem) = RecentDocuments.Where(Function(x) Not x.Pinned).OrderBy(Function(x) x.ItemDate)
                If notPinned.Count() > 0 Then
                    RecentDocuments.Remove(notPinned.Last())
                Else
                    RecentDocuments.Remove(RecentDocuments.OrderBy(Function(x) x.ItemDate).Last())
                End If
            End While
        End Sub
        ' get or set the list of recent documents.
        Public Property RecentDocuments As List(Of RecentDocumentItem)

        Public Sub RemoveDocument(ByVal fileName As String)
            Dim doc As RecentDocumentItem = RecentDocuments.FirstOrDefault(Function(x) x.FileName = fileName)
            If doc IsNot Nothing Then RecentDocuments.Remove(doc)
        End Sub
    End Class

    ''' <summary>
    ''' Represents each item in the most recently used documents list.
    ''' </summary>
    <Serializable>
    Public Class RecentDocumentItem

        Public Property Text As String

        Public Property SubText As String

        Public Property ItemDate As DateTime

        Public Property Pinned As Boolean

        Public ReadOnly Property FileName As String
            Get
                Return Path.Combine(SubText, Text)
            End Get
        End Property
    End Class
End Namespace

File: NetFramework/Ribbon/CS/WordPad/RecentDocumentList.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace WordPad
{
    /// <summary>
    /// Class that handles the recently used document list.
    /// Manages the list that appears in the application's main menu,
    /// including persistence to the settings file.
    /// </summary>
    public class RecentDocumentList
    {
        const int MAX_ITEMS = 10;

        // Initializes a new instance of a RecentDocumentList.
        public RecentDocumentList(List<RecentDocumentItem> docs)
        {
            RecentDocuments = docs;
        }

        // add a document to the MRU list
        public void AddDocument(string fileName)
        {
            // look for the file in the current list
            RecentDocumentItem doc = RecentDocuments.FirstOrDefault(x => x.FileName == fileName);
            if (doc != null)
            {
                // update date
                doc.Date = DateTime.Now;                
            }
            else if (!string.IsNullOrEmpty(fileName))
            {
                // not a member, add
                doc = new RecentDocumentItem
                {
                    Text = Path.GetFileName(fileName),
                    SubText = Path.GetDirectoryName(fileName),
                    Date = DateTime.Now
                };
                RecentDocuments.Add(doc);
            }

            // trim list  
            while (RecentDocuments.Count > MAX_ITEMS)
            {
                // first - check not pinned docs
                var notPinned = RecentDocuments.Where(x => !x.Pinned).OrderBy(x => x.Date);
                if (notPinned.Count() > 0)
                {
                    RecentDocuments.Remove(notPinned.Last());
                }
                else
                {
                    // remove last pinned doc
                    RecentDocuments.Remove(RecentDocuments.OrderBy(x => x.Date).Last());
                }                
            }
        }

        // get or set the list of recent documents.
        public List<RecentDocumentItem> RecentDocuments
        {
            get;
            set;
        }

        public void RemoveDocument(string fileName)
        {
            var doc = RecentDocuments.FirstOrDefault(x => x.FileName == fileName);
            if (doc != null)
                RecentDocuments.Remove(doc);
        }
    }

    /// <summary>
    /// Represents each item in the most recently used documents list.
    /// </summary>
    [Serializable]
    public class RecentDocumentItem
    {
        public string Text { get; set; }
        public string SubText { get; set; }
        public DateTime Date { get; set; }
        public bool Pinned { get; set; }     

        public string FileName
        {
            get { return Path.Combine(SubText, Text); }
        }       
    }
}
